src/Evolution/Systems/CurvedScalarWave/Worldtube/PunctureField.cpp

```diff
@@ -0,0 +1,322 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/CurvedScalarWave/Worldtube/PunctureField.hpp"
+
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/Systems/CurvedScalarWave/Tags.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace CurvedScalarWave::Worldtube {
+
+void compute_puncture_field(
+    gsl::not_null<Variables<tmpl::list<
+        Tags::Psi, ::Tags::dt<Tags::Psi>,
+        ::Tags::deriv<Tags::Psi, tmpl::size_t<3>, Frame::Inertial>>>*>
+        evolved_vars,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& coords, const double time,
+    const double orbital_radius, const double bh_mass, const size_t order) {
+  if (order == 0) {
+    compute_puncture_field_0(evolved_vars, coords, time, orbital_radius,
+                             bh_mass);
+  } else {
+    ERROR(
+        "Puncture order is only implemented up to order 0 but requested order "
+        << order);
+  }
+}
+
+void compute_puncture_field_0(
+    gsl::not_null<Variables<tmpl::list<
+        Tags::Psi, ::Tags::dt<Tags::Psi>,
+        ::Tags::deriv<Tags::Psi, tmpl::size_t<3>, Frame::Inertial>>>*>
+        evolved_vars,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& coords, const double time,
+    const double orbital_radius, const double BH_mass) {
+  evolved_vars->initialize(get<0>(coords).size());
+  const double r0 = orbital_radius;
+  const double M = BH_mass;
+  const double w = pow(r0, -1.5);
```

Comment:
not a full review, just a quick note. I realize most of this code probably comes from Mathematica, but assuming you'll be calling this quite frequently during the evolution you should go through and do optimizations like `pow(x, 2)` -> `pow<2>(x)`, or `pow(r0, -1.5)->1.0 / (r0 * sqrt(r0))` (pow is insanely expensive, you can do about 180 mult/add for one pow)